skglm/solvers/pn_solver.py

```diff
@@ -0,0 +1,295 @@
+from numba import njit
+
+from skglm.solvers.common import construct_grad, construct_grad_sparse
+
+
+def pn_solver(X, y, datafit, penalty, w_init=None, p0=10,
+              max_iter=20, max_epochs=1000, tol=1e-4, verbose=0):
+    """Run a Prox Newton solver.
+
+    Parameters
+    ----------
+    X : array, shape (n_samples, n_features)
+        Design matrix.
+
+    y : array, shape (n_samples,)
+        Target vector.
+
+    datafit : instance of BaseDatafit
+        Datafit object.
+
+    penalty : instance of BasePenalty
+        Penalty object.
+
+    w_init : array, shape (n_features,), default None
+        Initial value of coefficients.
+        If set to None, a zero vector is used instead.
+
+    p0 : int, default 10
+        Minimum number of groups to be included in the working set.
+
+    max_iter : int, default 20
+        Maximum number of iterations.
+
+    max_epochs : int, default 1000
+        Maximum number of epochs.
+
+    tol : float, default 1e-4
+        Tolerance for convergence.
+
+    verbose : bool, default False
+        Amount of verbosity. 0/False is silent.
+
+    Returns
+    -------
+    w : array, shape (n_features,)
+        Solution that minimizes the problem defined by datafit and penalty.
+
+    objs_out: array (max_iter,)
+        The objective values at every outer iteration.
+
+    stop_crit: float
+        The value of the stop criterion.
+    """
+    n_samples, n_features = X.shape
+    w = np.zeros(n_features) if w_init is None else w_init
+    Xw = np.zeros(n_samples) if w_init is None else X @ w_init
+    all_features = np.arange(n_features)
+    stop_crit = 0.
+    p_objs_out = []
+
+    is_sparse = issparse(X)
+    if is_sparse:
+        X_bundles = (X.data, X.indptr, X.indices)
+
+    for t in range(max_iter):
+        # compute scores
+        if is_sparse:
+            grad = construct_grad_sparse(*X_bundles, y, w, Xw, datafit, all_features)
+        else:
+            grad = construct_grad(X, y, w, Xw, datafit, all_features)
+
+        opt = penalty.subdiff_distance(w, grad, all_features)
+
+        # check convergences
+        stop_crit = np.max(opt)
+        if verbose:
+            p_obj = datafit.value(y, w, Xw) + penalty.value(w)
+            print(
+                f"Iteration {t+1}: {p_obj:.10f}, "
+                f"stopping crit: {stop_crit:.2e}"
+            )
+
+        if stop_crit <= tol:
+            if verbose:
+                print('Outer loop early exit')
+            break
+
+        # build working set
+        gsupp_size = penalty.generalized_support(w).sum()
+        ws_size = max(min(p0, n_features),
+                      min(n_features, 2 * gsupp_size))
+        # similar to np.argsort()[-ws_size:] but without sorting
+        ws = np.argpartition(opt, -ws_size)[-ws_size:]
+        grad_ws = grad[ws]
+
+        for epoch in range(max_epochs):
+            tol_in = 0.3 * stop_crit
+
+            # find descent direction
+            if is_sparse:
+                (
```

Comment:
cosmetic: `delta_w_ws, X_delta_w_ws = _compute_descent_direction(`  and break line there, right after the opening parenthesis